streams/src/main/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProvider.java

```diff
@@ -30,27 +32,35 @@
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
 
 public class StreamThreadStateStoreProvider {
 
     private final StreamThread streamThread;
+    private final InternalTopologyBuilder internalTopologyBuilder;
 
-    public StreamThreadStateStoreProvider(final StreamThread streamThread) {
+    public StreamThreadStateStoreProvider(final StreamThread streamThread, final InternalTopologyBuilder internalTopologyBuilder) {
         this.streamThread = streamThread;
+        this.internalTopologyBuilder = internalTopologyBuilder;
     }
 
     @SuppressWarnings("unchecked")
-    public <T> List<T> stores(final String storeName,
-                              final QueryableStoreType<T> queryableStoreType,
-                              final boolean includeStaleStores) {
+    public <T> List<T> stores(final StoreQueryParams storeQueryParams) {
+        final String storeName = storeQueryParams.getStoreName();
+        final QueryableStoreType<T> queryableStoreType = storeQueryParams.getQueryableStoreType();
+        final TaskId keyTaskId = createKeyTaskId(storeName, storeQueryParams.partition());
         if (streamThread.state() == StreamThread.State.DEAD) {
             return Collections.emptyList();
         }
         final StreamThread.State state = streamThread.state();
-        if (includeStaleStores ? state.isAlive() : state == StreamThread.State.RUNNING) {
-            final Map<TaskId, ? extends Task> tasks = includeStaleStores ? streamThread.allTasks() : streamThread.activeTasks();
+        if (storeQueryParams.staleStoresEnabled() ? state.isAlive() : state == StreamThread.State.RUNNING) {
+            final Map<TaskId, ? extends Task> tasks = storeQueryParams.staleStoresEnabled() ? streamThread.allTasks() : streamThread.activeTasks();
             final List<T> stores = new ArrayList<>();
             for (final Task streamTask : tasks.values()) {
+                if (keyTaskId != null && !keyTaskId.equals(streamTask.id())) {
+                    continue;
+                }
```

Comment:
Why do we have this check inside the loop? Should we not check this condition once, and only execute the loop if the condition is `false`?

Or maybe pass a `Collections.singleton(keyTaskId)` into the loop if `keyTaskId != null` ?

Reply:
Seems you missed this commnet? If you disagree with my comment, can you elaborate why (maybe I am missing something)?

Reply:
If keyTaskId is null(user wants all servable stores) -> loop over all tasks and return the list of stores.
If keyTaskId is not null(user wants a specific servable store) -> find the store from the specific task, by looping over all tasks and matching when keyTaskId matches the current task in loop.

So, probably I can write it as:
`if (keyTaskId != null) {
for (final Task streamTask : tasks.values()) {
if (keyTaskId.equals(streamTask.id()) { 
return streamTask.getStore(storeName); 
}
} else {
for (final Task streamTask : tasks.values()) {
//return list of stores
}
}`

But I feel the current code is more readable than this one.

Reply:
Why do loop over all tasks? If a user want to query a single partition, we can just lookup the `keyTaskId` in the map `tasks`:
```
if (keyTaskId != null) {
  return task.get(keyTaskId).getStore(storeName);
} else {
  for (final Task streamTask : tasks.values()) {
    //return list of stores
  }
}
```

Reply:
Done.